src/battlesnake.theserverproject.com/insert_clever_name/snake/engine/moveutils.go

```diff
@@ -0,0 +1,122 @@
+package engine
+
+import (
+	"battlesnake.theserverproject.com/insert_clever_name/snake/game"
+	"errors"
+)
+
+const (
+	UP    string    = "up"
+	DOWN  string    = "down"
+	LEFT  string    = "left"
+	RIGHT string    = "right"
+	OPPONENT string = "o"
+	NONE  string    = "n"
+)
+
+var MOVES = [...]string{UP, DOWN, LEFT, RIGHT}
+var OPPOSITE_MOVE = map[string]string{
+	UP:    DOWN,
+	DOWN:  UP,
+	LEFT:  RIGHT,
+	RIGHT: LEFT,
+}
+
+func checkMyMove(g game.Game, coord game.Coordinate) (bool, error) {
+	valid := true
+	value := g.Board[coord.Y][coord.X]
+	switch g.Board[coord.Y][coord.X] {
+	case game.EMPTY:
+		valid = true
+	case game.FOOD:
+		valid = true
+	case game.WALL:
+		valid = false
+	default:
+		if snakeValues, ok := g.ValueSnakeMap[value]; ok {
+			size := len(snakeValues.Body)
+			if coord.X == snakeValues.Body[size-1].X && coord.Y == snakeValues.Body[size-1].Y { // coord is a tail value and could be good or bad; evaluated during alpha-beta
+				valid = true
+			} else { // coord is a head or body value and means certain death
+				valid = false
+			}
+		} else {
+			return valid, errors.New("untracked board value")
+		}
+	}
+	return valid, nil
+}
+
+func prependHead(body []game.Coordinate, head game.Coordinate) []game.Coordinate { // theoretically faster than using straight append(), should be tested
+	body = append(body, head)
+	copy(body[1:], body)
+	body[0] = head
+	return body
+}
+
+func shiftBody(body []game.Coordinate, head game.Coordinate) []game.Coordinate {
+	for i := len(body) - 1; i >= 1; i-- {
+		body[i] = body[i - 1]
+	}
+	body[0] = head
+	return body
+}
+
+func moveMeTo(g game.Game, newHeadCoord game.Coordinate) *node {
```

Comment:
I think a lot of things may be simplified here if we move a snake based on its ID. I didn't do much looking at the code that already existed , but just did some digging and see that we track our own snake separately from opponent snakes. I think it would probably flow better for the `Game` struct to hold a `Me` ID rather than a `*SnakeValue`, then have all the snakes stored in the `SnakeValuesMap` (which should probably have a more descriptive name than [type]map, something like `LivingSnakes` or something along those lines (the fact that its a map can be seen by a developer, the purpose can't)). That way anything that needs to refer to all snakes can iterate over the `SnakeValuesMap`, and anything that needs a specific snake goes to `SnakeValuesMap` rather than arbitrarily either going to `Me` or `SnakeValuesMap`

Reply:
I'm not entirely sure what you mean, may have to talk for more elaboration. I changed the management of Snake information to reduce it to the minimum necessary. There aren't any SnakeValue maps anymore, just one ValueSnake mapping so that a snake's information can be accessed by its board value when we scan the board during evaluation and movement (ValueSnake[game.ME] is a unique case). I debated trying to make the move functions fit the cases for both us and the opponent snakes to reduce code duplication, but figured that the criteria of ruling out a move is different for us as it is for other snakes, so opted to keep the logic separate, at least for now.

Reply:
Okay, so I have left the moveMeTo and checkMyMove functions for reference, but added the beginnings of generic ones that _should_ be possible to reduce code duplication. I will remove these two functions when the generic ones are finished.

Reply:
Gotcha, I was probably looking at old code then for the namings.

Perfect, thank you I appreciate it!